[Feature] Support add multiple generated columns and reorder when adding multiple normal columns (#24885)

srlch opened the pull request:
1. ADD multiple generate column at once
```
ALTER TABLE t ADD COLUMN (newcol1 BIGINT AS id * 10, newcol2 BIGINT AS id * 100);

CREATE TABLE `t` (
  `id` bigint(20) NOT NULL COMMENT "",
  `newcol1` bigint(20) NULL AS `test`.`t`.`id` * 10 COMMENT "",
  `newcol2` bigint(20) NULL AS `test`.`t`.`id` * 100 COMMENT ""
) ENGINE=OLAP 
UNIQUE KEY(`id`)
DISTRIBUTED BY HASH(`id`) BUCKETS 7 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"compression" = "LZ4"
);
```

2. reorder normal column position when add multiple normal column into table with generated column
```
ALTER TABLE t ADD COLUMN (newcol3 BIGINT DEFAULT "0", newcol4 BIGINT DEFAULT "0");

CREATE TABLE `t` (
  `id` bigint(20) NOT NULL COMMENT "",
  `newcol3` bigint(20) NULL DEFAULT "0" COMMENT "",
  `newcol4` bigint(20) NULL DEFAULT "0" COMMENT "",
  `newcol1` bigint(20) NULL AS `test`.`t`.`id` * 10 COMMENT "",
  `newcol2` bigint(20) NULL AS `test`.`t`.`id` * 100 COMMENT ""
) ENGINE=OLAP 
UNIQUE KEY(`id`)
DISTRIBUTED BY HASH(`id`) BUCKETS 7 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"compression" = "LZ4"
);
```

Fixes #24885

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1 
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 71 (21.13%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|7|00.00%|[179, 180, 181, 183, 185, 186, 187]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|7|45|15.56%|[409, 411, 412, 415, 416, 418, 419, 421, 424, 425, 426, 427, 428, 430, 433, 434, 435, 436, 437, 438, 439, 441, 442, 444, 447, 448, 449, 450, 452, 458, 462, 463, 464, 465, 466, 467, 468, 469]|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnsClause.java|1|4|25.00%|[32, 36, 37]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|15|46.67%|[3590, 3591, 3592, 3593, 3596, 3597, 3598, 3599]|

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java:
```diff
@@ -393,6 +393,8 @@ public Void visitAddColumnsClause(AddColumnsClause clause, ConnectContext contex
             if (columnDefs == null || columnDefs.isEmpty()) {
                 throw new SemanticException("Columns is empty in add columns clause.");
             }
+            boolean hasMaterializedColumn = false;
```
You should change all the `materializedColumn` to `generatedColumn` in the codepaths.

chaoyli commented on test/sql/test_materialized_column/R/test_materialized_column:
```diff
@@ -518,3 +518,87 @@ DROP TABLE t;
+-- !result
+ALTER TABLE t ADD COLUMN (newcol5 BIGINT DEFAULT "0", newcol6 BIGINT AS id * 1000);
+-- result:
+E: (1064, 'Getting analyzing error. Detail message: Can not add normal column and generated column in the same time.')
```
```
If we add multiple normal columns into a table with generated columns, the cols will be located after generated column.
```
It seems not consistent with the needs you describe.

srlch replied:
I have update the issue message to be more clear. And the example for the sql-test result is the limitation for adding multiple column: we can add multiple normal column and also add  multiple generated column at once , but we can not add mix normal column and generated column at the same time for simplicity

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 13 / 69 (18.84%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|7|00.00%|[179, 180, 181, 183, 185, 186, 187]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|7|45|15.56%|[409, 411, 412, 415, 416, 418, 419, 421, 424, 425, 426, 427, 428, 430, 433, 434, 435, 436, 437, 438, 439, 441, 442, 444, 447, 448, 449, 450, 452, 458, 462, 463, 464, 465, 466, 467, 468, 469]|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnsClause.java|1|4|25.00%|[32, 36, 37]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|13|38.46%|[3590, 3591, 3592, 3593, 3596, 3597, 3598, 3599]|

srlch committed:
[Feature] Support add multiple generated column and reorder when adding normal column (#24885)

This pr solve two problem:

Problem 1:
Currently, if we add a normal column into a table with generated column. The normal column will
be added automatically before the generated column. We will keep the order of the columns that
generated columns must be at the last. But it is not work if we add multiple normal column at
once. The pr fix the problem.

Problem 2:
Support add multiple generated column using at once as following:
ALTER TABLE t ADD COLUMN (newcol1 INT AS id, newcol2 as id * 10)

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 13 / 69 (18.84%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|7|00.00%|[179, 180, 181, 183, 185, 186, 187]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|7|45|15.56%|[409, 411, 412, 415, 416, 418, 419, 421, 424, 425, 426, 427, 428, 430, 433, 434, 435, 436, 437, 438, 439, 441, 442, 444, 447, 448, 449, 450, 452, 458, 462, 463, 464, 465, 466, 467, 468, 469]|
|:large_blue_circle:|com/starrocks/sql/ast/AddColumnsClause.java|1|4|25.00%|[32, 36, 37]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|13|38.46%|[3590, 3591, 3592, 3593, 3596, 3597, 3598, 3599]|
